Sources/_RegexParser/Regex/Parse/Sema.swift

```diff
@@ -0,0 +1,396 @@
+    }
+  }
+
+  func validateMatchingOption(_ opt: AST.MatchingOption) throws {
+    let loc = opt.location
+    switch opt.kind {
+    case .allowDuplicateGroupNames:
+      // Not currently supported as we need to figure out what to do with
+      // the capture type.
+      throw error(.unsupported("duplicate group naming"), at: loc)
+
+    case .unicodeWordBoundaries:
+      throw error(.unsupported("unicode word boundary mode"), at: loc)
+
+    case .textSegmentWordMode, .textSegmentGraphemeMode:
+      throw error(.unsupported("text segment mode"), at: loc)
+
+    case .byteSemantics:
+      throw error(.unsupported("byte semantic mode"), at: loc)
+
+    case .caseInsensitive, .possessiveByDefault, .reluctantByDefault,
+        .unicodeScalarSemantics, .graphemeClusterSemantics,
+        .singleLine, .multiline, .namedCapturesOnly, .extended, .extraExtended,
+        .asciiOnlyDigit, .asciiOnlyWord, .asciiOnlySpace, .asciiOnlyPOSIXProps:
+      break
+    }
+  }
+
+  func validateMatchingOptions(_ opts: AST.MatchingOptionSequence) throws {
+    for opt in opts.adding {
+      try validateMatchingOption(opt)
+    }
+    for opt in opts.removing {
+      try validateMatchingOption(opt)
+    }
+  }
+
+  func validateBinaryProperty(
+    _ prop: Unicode.BinaryProperty, at loc: SourceLocation
+  ) throws {
+    switch prop {
+    case .asciiHexDigit, .alphabetic, .bidiMirrored, .cased, .caseIgnorable,
+        .changesWhenCasefolded, .changesWhenCasemapped,
+        .changesWhenNFKCCasefolded, .changesWhenLowercased,
+        .changesWhenTitlecased, .changesWhenUppercased, .dash, .deprecated,
+        .defaultIgnorableCodePoint, .diacratic, .extender,
+        .fullCompositionExclusion, .graphemeBase, .graphemeExtended, .hexDigit,
+        .idContinue, .ideographic, .idStart, .idsBinaryOperator,
+        .idsTrinaryOperator, .joinControl, .logicalOrderException, .lowercase,
+        .math, .noncharacterCodePoint, .patternSyntax, .patternWhitespace,
+        .quotationMark, .radical, .regionalIndicator, .softDotted,
+        .sentenceTerminal, .terminalPunctuation, .unifiedIdiograph, .uppercase,
+        .variationSelector, .whitespace, .xidContinue, .xidStart:
+      break
+
+    case .emojiModifierBase, .emojiModifier, .emoji, .emojiPresentation:
+      // These are available on macOS 10.12.2, iOS 10.2, tvOS 10.1, watchOS 3.1.1.
+      // TODO: We should ideally check deployment target for such conditionally
+      // available properties.
+      break
+
+    case .expandsOnNFC, .expandsOnNFD, .expandsOnNFKD, .expandsOnNFKC:
+      throw error(.deprecatedUnicode(prop.rawValue.quoted), at: loc)
+
+    case .bidiControl, .compositionExclusion, .emojiComponent,
+        .extendedPictographic, .graphemeLink, .hyphen, .otherAlphabetic,
+        .otherDefaultIgnorableCodePoint, .otherGraphemeExtended,
+        .otherIDContinue, .otherIDStart, .otherLowercase, .otherMath,
+        .otherUppercase, .prependedConcatenationMark:
```

Comment:
For these `.other*` properties, we don't implement b/c they're included in e.g. `\p{isAlphabetic}`. Do you think we should redirect people to the corresponding properties?

Reply:
Makes sense, and I can take care of that in a follow-up